### DISCUSSION TITLE
Add possibility of divisor spanning zero

### DISCUSSION THREAD

**Discussion by snedeljkovic** [google/or-tools#3207] [Feature requests] [answered]
**What language and solver does this apply to?**
Python, CP-SAT

**Describe the problem you are trying to solve.**
I'm modelling abstract syntax trees for symbolic computation.

**Describe the solution you'd like**
As in the name of the request I'd like the possibility of the divisor to span across zero.

**Describe alternatives you've considered**

**Additional context**

**Comment by lperron** [comment, marked as answer]
Noted.

It will not happen soon.  

You can emulate it by looking at: https://github.com/google/or-tools/blob/b37d9c786b69128f3505f15beca09e89bf078a89/ortools/sat/cp_model_expand.cc#L255

Roughly speaking, 

divide the domain of the divisor in 2 (< 0, >0). Create one BoolVar b, link it to the 2 side of the domains.
Create 2 temp divisor variables (pos_div, neg_div), 2 temp target variable (pos_target, neg_target).

```
b implies divisor > 0
b implies pos_divisor == divisor
b implies neg_divisor == -1  # or max of the negative part of the divisor domain
add_division_equality(pos_target, numerator, pos_divisor)

not(b) implies divisor < 0
not(b) implies neg_divisor == divisor
not(b) implies pos_target == 1  # or min of the positive part of the divisor domain
add_division_equality(neg_target, numerator, neg_divisor)

link it all together

b implies target == pos_target
not(b) implies target == neg_target

**Reply by snedeljkovic** [reply]
Thanks, Laurent!

**Comment by lperron** [comment]
I have just pushed the code for that.

Never too late :-)
